### ISSUE TITLE
Use predefined name for the ligature parts

### ISSUE THREAD

**Issue by leiserfg** [be5invis/Iosevka#1007]
Hi @be5invis, have you checked this comment from @kovidgoyal?
https://github.com/kovidgoyal/kitty/issues/297#issuecomment-779134244
It will be pretty nice to have the best monospaced font and the best terminal emulator working together :pray:

**Comment by be5invis** [comment]
Kitty should fix their broken text layout, **even if it slows down everything**, instead of asking font authors to compromise.

**Comment by kovidgoyal** [comment]
How is naming your glyphs sensibly a compromise? You want to do stupid ass things like have infinite length ligatures, at least do it right.

**Comment by be5invis** [comment]
@kovidgoyal If we do this, perhaps Kitty could support `.init`/`.medi`/`.fina`/`.isol` too. `.init`/`.medi`/`.fina`/`.isol` are how font makers usually joining letterforms, like that in Arabic.

**Comment by kovidgoyal** [comment]
Is there some reason you want to use a different naming scheme
from other fonts that support infinite length ligatures? It would
be nice to have a consistent naming scheme across fonts.

To be precise, glyphs that are part of an infinite ligature are
currently assumed to have the following naming scheme

prefix_(start|middle|end).seq

where prefix is must not contain the _ character. This scheme
is followed by at least Fira Code and Cascadia Code and possibly
MonoLisa, though am not sure of the last.

If you have some reason to want a different naming scheme, I am willing
to implement support for it, though I would urge you to keep it as
simple as possible to avoid making lookups unnecessarily expensive.

**Comment by kovidgoyal** [comment]
Here is the function used to detect infinite ligature glyphs in kitty: https://github.com/kovidgoyal/kitty/blob/master/kitty/fonts.c#L810

**Comment by be5invis** [comment]
@kovidgoyal One problem with your pattern is that all ligations must follow the "start-middle-end" pattern, while in Iosevka, the glyphs are used more freely (for example, ligation of `!=` is actually an equal symbol with a bar overlay.

And sometimes ligations will interact with each other, so making ligations' names follow this pattern will further complicate implementations.

![image](https://user-images.githubusercontent.com/240091/118599071-da79e900-b763-11eb-8417-0224256c8619.png)

I am curious about whether it is possible to have a _single_ suffix to indicate that this glyph is a part of a ligature. Or perhaps we could leverage [GDEF's Glyph Class Definition](https://docs.microsoft.com/en-us/typography/opentype/spec/gdef#glyph-class-definition-table). Once Kitty identified a continuous sequence of glyphs are all considered "part of ligature", it will render them together.

**Comment by kovidgoyal** [comment]
On Mon, May 17, 2021 at 11:05:42PM -0700, Belleve wrote:
> @kovidgoyal One problem with your pattern is that all ligations must follow the "start-middle-end" pattern, while in Iosevka, the glyphs are used more freely (for example, ligation of `!=` is actually an equal symbol with a bar overlay.

It isn't my pattern, it's what the fonts I mentioned earlier use. I just
added code to kitty to take advantage of it. Note that it is needed only
for infinite length ligatures, not normal ones like != or === or => etc
These types of ligatures are represented by either:

1) A single ligature glyph
2) One or more empty glyphs + a ligature glyph
3) A ligature glyph + one or more empty glyphs

kitty handles these cases already, without needing to check 
the glyph name. 

It is only when the ligature consists of N non empty glyphs that it
becomes tricky to detect the boundaries of the ligature and therefore
interrogating the name becomes necessary. In practice this only happens
for infinite length ligatures.

> 
> And sometimes ligations will interact with each other, so making ligations' names follow this pattern will further complicate implementations.

Not sure I follow. ligature glyph names are currently arbitrary, often
something like gid<glyph number>. Imposing additional structure on them
canonly make things better.

> 
> ![image](https://user-images.githubusercontent.com/240091/118599071-da79e900-b763-11eb-8417-0224256c8619.png)
> 
> I am curious about whether it is possible to have a _single_ suffix to indicate that this glyph is a part of a ligature.

It's not possible. Think of two ligatures that are logically separate,
but neighboring. If all the glyphs have the same suffix, then there is
no reasonable way to tell the end of one and the start of the next apart. That's
why one needs the start/middle/end triplet.

> Or perhaps we could leverage [GDEF's Glyph Class Definition](https://docs.microsoft.com/en-us/typography/opentype/spec/gdef#glyph-class-definition-table). Once Kitty identified a continuous sequence of glyphs are all considered "part of ligature", it will render them together.

Same issue as with single suffix, I think.

**Comment by be5invis** [comment]
> It's not possible. Think of two ligatures that are logically separate,
> but neighboring. If all the glyphs have the same suffix, then there is
> no reasonable way to tell the end of one and the start of the next apart. 

Do we need to handle this case separately? If it is rendering-only then combining adjacent ligatures into a long run won't influence much, even for performance, since ligatures do not usually meet.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 12:03:28AM -0700, Belleve wrote:
> > It's not possible. Think of two ligatures that are logically separate,
> > but neighboring. If all the glyphs have the same suffix, then there is
> > no reasonable way to tell the end of one and the start of the next apart. 
> 
> Do we need to handle this case separately? If it is rendering-only then combining adjacent ligatures into a long run won't influence much, even for performance, since ligatures do not usually meet.

>From a purely rendering based perspective, it will *likely* work fine
(the devil, as always is in the details, there may be issues with
leading/kerning in practice). 

However, it is desirable to be able to tell logical glyphs apart, for
many reasons. For instance, kitty has the feature that you can set an
option so that when the cursor is over a CALT ligature it is rendered
normally, not as a ligature. This makes editing it easier. If neighboring
ligatures all become one large ligature, it would mean this feature
would no longer work well. 

And then there is performance, having unnecessarily large ligatures
hurts performance, and at least in current kitty, where there is a 9
character maximum size limit on ligatures, neighboring ligatures could
easily go over that mlimit when individually all of them would be fine.

**Comment by be5invis** [comment]
I do not think neighboring ligatures occur often in real life — for code usages, they are usually operators separated by spaces, and for terminal output, I do not think they are frequent too.

To clarify: if we need to implement the `start`/`middle`/`end` glyph name suffixes I need to rewrite **almost all code related to ligation**, and it may introduce a lot more doppelganger glyphs (glyphs having the same geometry but different GID/name). Also, it will make _shaping_ slower since we need more OTL lookups to "correct" the glyphs to the properly named one.

For example, the image below shows `<~~~~~~~~`, and the long tail is actually built up with a single glyph that _overlays with the glyph before it_. Implementing such a tail will need only one lookup (`sub tail_end tilde' by tail_end`), but if we want the glyphs to follow the naming pattern, we will need an extra lookup to "fix" the waves in between (like `sub tail_end' tail_end by tail_mid`).

![image](https://user-images.githubusercontent.com/240091/118610921-23846a00-b771-11eb-9267-6ba500ed7795.png)

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 12:41:06AM -0700, Belleve wrote:
> I do not think neighboring ligatures occur often in real life — for code usages, they are usually operators separated by spaces, and for terminal output, I do not think they are frequent too.

Operators often dont have spaces around them.

> 
> To clarify: if we need to implement the `start`/`middle`/`end` glyph name suffixes I need to rewrite **almost all code related to ligation**, and it may introduce a lot more doppelganger glyphs (glyphs having the same geometry but different GID/name). Also, it will make _shaping_ slower since we need more OTL lookups to "correct" the glyphs to the properly named one.

You dont need middle, just start and end, middle is only needed if there
are more than two glyphs.

> 
> For example, the image below shows `<~~~~~~~~`, and the long tail is actually built up with a single glyph that _overlays with the glyph before it_. Implementing such a tail will need only one lookup (`sub tail_end tilde' by tail_end`), but if we want the glyphs to follow the naming pattern, we will need an extra lookup to "fix" the waves in between (like `sub tail_end' tail_end by tail_mid`).

See above, you still need only two glyphs, just the first one should
have a name ending with _start.seq and the second a name ending with 
_end.seq

**Comment by be5invis** [comment]
> See above, you still need only two glyphs, just the first one should have a name ending with _start.seq and the second a name ending with _end.seq

So your implant supports “start-end-end…-end” and “start-start-start-…-end” pattern?

Or the “start” means “this glyph must be rendered with the glyph right after it”, while “end” means “this glyph must be rendered with the glyph right before it”?

**Comment by kovidgoyal** [comment]
kitty will recognize all of the following:

1) start, end
2) start, middle, ... , middle, end
3) middle, end
4) middle, middle, ... , middle, end

**Comment by be5invis** [comment]
I'd like to have a behavior that:

 * If a glyph's name ends with `_start.seq`, then it must be rendered with the (non-mark) glyph right after it, no matter what it is, or how it is named;
 * If a glyph's name ends with `_end.seq`, then it must be rendered with the (non-mark) glyph right before it, no matter what it is, or how it is named;
 * If a glyph's name ends with `_middle.seq`, then it must be rendered with the (non-mark) glyph right after and right before it, no matter what they are, or how they are named.

This will allow patterns like `start--start--start--end`, `start--end--end--end` or even `isolate--end--middle--end`. This will simplify the work of font designers a lot.

**Comment by be5invis** [comment]
| Before\\After | Isolated | Start    | Middle | End  |
| ------------- | -------- | -------- | ------ | ---- |
| Isolate       | Separate | Separate | Join   | Join |
| Start         | Join     | Join     | Join   | Join |
| Middle        | Join     | Join     | Join   | Join |
| End           | Separate | Separate | Join   | Join |

**Comment by kovidgoyal** [comment]
er to use a
different naming scheme, because those rules will likely break other fonts. In
particular the requirement that end/middle glyphs always force joins. 

So I suggest picking a different suffix instead of .seq, maybe .calt-lig or
even .iosevka.

Then I will have kitty check the glyphs for the === ligature. If one of them
has the suffix .iosevka it will use the rules you want, otherwise it will
use the current FiraCode/CascadiaCode rules.

Also, if you are going to setup such rules, probably would be good to use them
for all ligatures not just infinite length ones. That will make the code to
recognize ligatures much simpler and more robust and will probably make your
code to generate them also simpler and more robust.

**Comment by leiserfg** [comment]
I think if that's the way, maybe a more generic suffix (not iosevka) will be better, that way other font designers could be able to use it too (like .liga, .calt-lig,  .kitty). And maybe it can be exposed as somehow in the font  (I'm not sure if fonts can have custom properties)  so the `===` heuristic is not required, that way it could also be used for non-programming fonts.

**Comment by kovidgoyal** [comment]
Yes, better to make it into a general specification that can be widely used, but that requires more work :)

**Comment by be5invis** [comment]
@leiserfg @kovidgoyal I am thinking if we need to implement a joining-based semantics, the glyph naming pattern could be `.join-r`, `.join-l` and `.join-m`, indicating whether a glyph will join with the adjacent glyph at right, at left, or at both left and right.

The table is:

| Before\\After | Isolated | `.join-r` | `.join-m` | `.join-l` |
| ------------- | -------- | -------- | ------ | ---- |
| Isolated      | Separate | Separate | Join   | Join |
| `.join-r`     | Join     | Join     | Join   | Join |
| `.join-m`     | Join     | Join     | Join   | Join |
| `.join-l`     | Separate | Separate | Join   | Join |

**Comment by kovidgoyal** [comment]
Fine by me, though since the direction of text can be LTR or RTL
probably better to use p (previous) n (next) and b (both)

**Comment by be5invis** [comment]
@kovidgoyal Well, I think `l` and `r` are better for font designers, since glyph visuals are, well, visual.

**Comment by be5invis** [comment]
One sidenote: the joining-ness between glyph pair is sort of disjunction, i.e., if the glyph at left has `.join-r` while the glyph at right is not considered joining, **Kitty should still glue them together**, because the first glyph indicates it must join with the glyph after it. The table above shows all the cases of adjacent glyph pairs.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 06:59:03PM -0700, Belleve wrote:
> @kovidgoyal Well, I think `l` and `r` are better for font designers, since glyph visuals are, well, visual.

Suit yourself, it doesnt really matter to me.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 07:01:31PM -0700, Belleve wrote:
> One sidenote: the joining-ness between glyph pair is sort of disjunction, i.e., if the glyph at left has `.join-r` while the glyph at right is not considered joining, **Kitty should still glue them together**, because the first glyph indicates it must join with the glyph after it. The table above shows all the cases of adjacent glyph pairs.

Yes, this is one of the reasons your scheme will break the other fonts.

In your scheme a ligature starts with either:

a) a join-r
b) a non-join followed by a join-l or a join-m 

Ends with only:
a join-l or join-m followed by a non-join.

**Comment by be5invis** [comment]
@kovidgoyal I think the join-suffix and seq-suffix are mutually exclusive, i.e., a font designer won't mix two kinds of suffixes in one project.

> In your scheme a ligature starts with either:

> a) a join-r
> b) a non-join followed by a join-l or a join-m

> Ends with only:
> a join-l or join-m followed by a non-join. 

The end condition should be 
1. a `.join-l`.
2. Or a non-join, following a `join-m` or `join-r`.

**Comment by be5invis** [comment]
Start condition: A `.join-r` or non-joining, **NOT** after a `.join-r` or `.join-m`;
End condition: A `.join-l` or non-join, **NOT** preceding a `.join-l` or `.join-m`.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 07:16:32PM -0700, Belleve wrote:
> @kovidgoyal I think the join-suffix and seq-suffix are mutually exclusive, i.e., a font designer won't mix two kinds of suffixes in one project.

Yes, most definitely.

> 
> > In your scheme a ligature starts with either:
> 
> > a) a join-r
> > b) a non-join followed by a join-l or a join-m
> 
> > Ends with only:
> > a join-l or join-m followed by a non-join. 
> 
> The end condition should be 
> 1. a `.join-l`.

What if a join-l is followed by another join-l or a join-m?

**Comment by be5invis** [comment]
> What if a join-l is followed by another join-l or a join-m?

The ligature will continue. I think you are using mail lost some of my edits, here is the revised conditions of whether a glyph starts or ends a ligature:

Start condition: Either
 - A `.join-r`, **NOT** after a `.join-r` or `.join-m`;
 - A non-joining, preceding a `.join-l` or `.join-m`, and **NOT** after a `.join-r` or `.join-m`;

End condition: Either
 - A `.join-l` **NOT** preceding a `.join-l` or `.join-m`.
 - A non-joining, after a `.join-r` or `.join-m`, and **NOT** preceding a `.join-l` or `.join-m`.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 07:19:57PM -0700, Belleve wrote:
> Start condition: A `.join-r` or non-joining, **NOT** after a `.join-r` or `.join-m`;

That start condition cant be correct. It means two non-joining glyphs
are a ligature. I think it would be easiest to just list all pairs of
glyphs you need as start conditions and all pairs as end conditions,
that way it is unambiguous.

**Comment by be5invis** [comment]
> It means two non-joining glyphs are a ligature.

For non-joining glyphs, it can start a glyph if and only it is right before a (`.join-l` or `.join-m`) glyph, and is **NOT** after a (`.join-r` or `.join-m`) glyph. So two adjacent non-joining glyph won't start a ligature.

**Comment by be5invis** [comment]
I think I need to add brackets:

Start condition: Either
 - A `.join-r`, **NOT** after a (`.join-r` or `.join-m`);
 - A non-joining, preceding a (`.join-l` or `.join-m`), and **NOT** after a (`.join-r` or `.join-m`);

End condition: Either
 - A `.join-l` **NOT** preceding a (`.join-l` or `.join-m`).
 - A non-joining, after a (`.join-r` or `.join-m`), and **NOT** preceding a (`.join-l` or `.join-m`).

**Comment by be5invis** [comment]
Pseudocode:

```javascript
function glyphShouldStartLigature(glyphBefore, glyphCurrent, glyphAfter) {
	return (
		(glyphCurrent.joining == R ||
			(!glyphCurrent.joining && (glyphAfter.joining == L || glyphAfter.joining == M))) &&
		!(glyphBefore.joining == R || glyphBefore.joining == M)
	);
}
function glyphShouldEndLigature(glyphBefore, glyphCurrent, glyphAfter) {
	return (
		(glyphCurrent.joining == L ||
			(!glyphCurrent.joining && (glyphBefore.joining == R || glyphBefore.joining == M))) &&
		!(glyphAfter.joining == L || glyphAfter.joining == M)
	);
}
```

**Comment by kovidgoyal** [comment]
OK, I will ananlyse it for conflicts in a while.

**Comment by be5invis** [comment]
The table below enumerates all possible combinations of pre-current-post joining classes: brackets indicating glyphs, hyphens in bracket shows the side this glyph needs to glue with, and hyphens between brackets indicating that these two adjacent glyphs will glue together. Hope it helps understanding how the joining classes work.

```
[ I ] [ I ] [ I ]    [ I ] [ I ] [ R-]    [ I ] [ I ]-[-L ]    [ I ] [ I ]-[-I-]
[ R-]-[ I ] [ I ]    [ R-]-[ I ] [ R-]    [ R-]-[ I ]-[-L ]    [ R-]-[ I ]-[-I-]
[-L ] [ I ] [ I ]    [-L ] [ I ] [ R-]    [-L ] [ I ]-[-L ]    [-L ] [ I ]-[-I-]
[-I-]-[ I ] [ I ]    [-I-]-[ I ] [ R-]    [-I-]-[ I ]-[-L ]    [-I-]-[ I ]-[-I-]
[ I ] [ R-]-[ I ]    [ I ] [ R-]-[ R-]    [ I ] [ R-]-[-L ]    [ I ] [ R-]-[-I-]
[ R-]-[ R-]-[ I ]    [ R-]-[ R-]-[ R-]    [ R-]-[ R-]-[-L ]    [ R-]-[ R-]-[-I-]
[-L ] [ R-]-[ I ]    [-L ] [ R-]-[ R-]    [-L ] [ R-]-[-L ]    [-L ] [ R-]-[-I-]
[-I-]-[ R-]-[ I ]    [-I-]-[ R-]-[ R-]    [-I-]-[ R-]-[-L ]    [-I-]-[ R-]-[-I-]
[ I ]-[-L ] [ I ]    [ I ]-[-L ] [ R-]    [ I ]-[-L ]-[-L ]    [ I ]-[-L ]-[-I-]
[ R-]-[-L ] [ I ]    [ R-]-[-L ] [ R-]    [ R-]-[-L ]-[-L ]    [ R-]-[-L ]-[-I-]
[-L ]-[-L ] [ I ]    [-L ]-[-L ] [ R-]    [-L ]-[-L ]-[-L ]    [-L ]-[-L ]-[-I-]
[-I-]-[-L ] [ I ]    [-I-]-[-L ] [ R-]    [-I-]-[-L ]-[-L ]    [-I-]-[-L ]-[-I-]
[ I ]-[-I-]-[ I ]    [ I ]-[-I-]-[ R-]    [ I ]-[-I-]-[-L ]    [ I ]-[-I-]-[-I-]
[ R-]-[-I-]-[ I ]    [ R-]-[-I-]-[ R-]    [ R-]-[-I-]-[-L ]    [ R-]-[-I-]-[-I-]
[-L ]-[-I-]-[ I ]    [-L ]-[-I-]-[ R-]    [-L ]-[-I-]-[-L ]    [-L ]-[-I-]-[-I-]
[-I-]-[-I-]-[ I ]    [-I-]-[-I-]-[ R-]    [-I-]-[-I-]-[-L ]    [-I-]-[-I-]-[-I-]
```

**Comment by be5invis** [comment]
The table below enumerates all possible combinations of pre-current-post joining classes: brackets indicating glyphs, hyphens in bracket shows the side this glyph needs to glue with, and hyphens between brackets indicating that these two adjacent glyphs will glue together. Hope it helps understanding how the joining classes work.

```
[ I ] [ I ] [ I ]    [ I ] [ I ] [ R-]    [ I ] [ I ]-[-L ]    [ I ] [ I ]-[-M-]
[ R-]-[ I ] [ I ]    [ R-]-[ I ] [ R-]    [ R-]-[ I ]-[-L ]    [ R-]-[ I ]-[-M-]
[-L ] [ I ] [ I ]    [-L ] [ I ] [ R-]    [-L ] [ I ]-[-L ]    [-L ] [ I ]-[-M-]
[-M-]-[ I ] [ I ]    [-M-]-[ I ] [ R-]    [-M-]-[ I ]-[-L ]    [-M-]-[ I ]-[-M-]
[ I ] [ R-]-[ I ]    [ I ] [ R-]-[ R-]    [ I ] [ R-]-[-L ]    [ I ] [ R-]-[-M-]
[ R-]-[ R-]-[ I ]    [ R-]-[ R-]-[ R-]    [ R-]-[ R-]-[-L ]    [ R-]-[ R-]-[-M-]
[-L ] [ R-]-[ I ]    [-L ] [ R-]-[ R-]    [-L ] [ R-]-[-L ]    [-L ] [ R-]-[-M-]
[-M-]-[ R-]-[ I ]    [-M-]-[ R-]-[ R-]    [-M-]-[ R-]-[-L ]    [-M-]-[ R-]-[-M-]
[ I ]-[-L ] [ I ]    [ I ]-[-L ] [ R-]    [ I ]-[-L ]-[-L ]    [ I ]-[-L ]-[-M-]
[ R-]-[-L ] [ I ]    [ R-]-[-L ] [ R-]    [ R-]-[-L ]-[-L ]    [ R-]-[-L ]-[-M-]
[-L ]-[-L ] [ I ]    [-L ]-[-L ] [ R-]    [-L ]-[-L ]-[-L ]    [-L ]-[-L ]-[-M-]
[-M-]-[-L ] [ I ]    [-M-]-[-L ] [ R-]    [-M-]-[-L ]-[-L ]    [-M-]-[-L ]-[-M-]
[ I ]-[-M-]-[ I ]    [ I ]-[-M-]-[ R-]    [ I ]-[-M-]-[-L ]    [ I ]-[-M-]-[-M-]
[ R-]-[-M-]-[ I ]    [ R-]-[-M-]-[ R-]    [ R-]-[-M-]-[-L ]    [ R-]-[-M-]-[-M-]
[-L ]-[-M-]-[ I ]    [-L ]-[-M-]-[ R-]    [-L ]-[-M-]-[-L ]    [-L ]-[-M-]-[-M-]
[-M-]-[-M-]-[ I ]    [-M-]-[-M-]-[ R-]    [-M-]-[-M-]-[-L ]    [-M-]-[-M-]-[-M-]
```

**Comment by kovidgoyal** [comment]
Looks fine, remaining question is how to detect the use of this scheme. My original proposal is to check for the suffix in one of the suffix in one of the glyphs of the === ligature, if you have a better proposal, let me know.

**Comment by be5invis** [comment]
@kovidgoyal You can test `==`, it is in the common set that all the ligation sets that Iosevka has.

**Comment by kovidgoyal** [comment]
OK, thanks.

**Comment by be5invis** [comment]
@kovidgoyal I added this to the backlog -- I think it could arrive in 7.0 or 7.1, depending on how much bandwidth I have :).

**Comment by be5invis** [comment]
@kovidgoyal Also, if you plan to implement this feature now, allow names like `some.name.with.dots.join-l`, i.e., dots before `.join-l`/`.join-m`/`.join-r`.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 08:32:40PM -0700, Belleve wrote:
> @kovidgoyal I added this to the backlog -- I think it could arrive in 7.0 or 7.1, depending on how much bandwidth I have :).

OK it should be the work of a couple of hours to add support for it in
kitty once I have a build of Iosevka to test with.

**Comment by be5invis** [comment]
@kovidgoyal 🍻 I'll notify you when I have files with these so you could test against it.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 08:35:00PM -0700, Belleve wrote:
> @kovidgoyal Also, if you plan to implement this feature now, allow names like `some.name.with.dots.join-l`, i.e., dots before `.join-l`/`.join-m`/`.join-r`.

OK, no problem.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 08:38:55PM -0700, Belleve wrote:
> @kovidgoyal 🍻 I'll notify you when I have files with these so you could test against it.

Cool, will be good to put this issue to bed :)
Thanks for the collaboration.

**Comment by kovidgoyal** [comment]
I had a little free tie so I implemented it: https://github.com/kovidgoyal/kitty/commit/46a0566e2e9b1ac0c5ec6e0e1a2cf726e45212f0  will add tests and finish the implementation once you send me a build.

**Comment by be5invis** [comment]
@kovidgoyal Thanks, I will notify you when I have one -- perhaps next week or early-mid June.

**Comment by be5invis** [comment]
Hi @kovidgoyal 
I have an internal build with name suffixes. Could you please try it in Kitty to see whether ligations are properly linked?

Test corpus:
```
-<< -< -<- <-- <--- <<- <- -> ->> --> ---> ->- >- >>-
=<< =< =<= <== <=== <<= <= => =>> ==> ===> =>= >= >>=
<-> <--> <---> <----> <=> <==> <===> <====> -------->
<~~ <~ ~> ~~> :: ::: == != /= ~= <> === !== =/= =!=
:= :- :+ <* <*> *> <| <|> |> <. <.> .> +: -: =: :> __
(* *) [| |] {| |} ++ +++ \/ /\ |- -| <!-- <!--- <***>
˩˥ ˧˥ ˩˦˩˧ ꜒꜒ ꜒꜓ ꜔꜕꜓꜖
```

Please test against `calt` and also `dlig` if possible.

[iosevka-regular.zip](https://github.com/be5invis/Iosevka/files/6582946/iosevka-regular.zip)

**Comment by kovidgoyal** [comment]
Thanks, will do so this weekend.

**Comment by kovidgoyal** [comment]
Is there a reference for what they are supposed to look like? I have attached a screenshot in kitty running from master, most ofthem look correct to me, except for some of the ones in the first and second row.
![ligs](https://user-images.githubusercontent.com/1308621/120671160-bc73ee80-c4ae-11eb-8e25-07c8fad47321.png)

**Comment by leiserfg** [comment]
@kovidgoyal  that's with just `calt` isn't, those that look rare use `dlig` I guess.

https://raw.githubusercontent.com/be5invis/Iosevka/master/images/ligset-calt-1.png
https://raw.githubusercontent.com/be5invis/Iosevka/master/images/ligset-dlig-1.png

**Comment by leiserfg** [comment]
This is how it looks in Wezterm
![image](https://user-images.githubusercontent.com/2947276/120701502-3312eb00-c4b3-11eb-9f79-0113d91de9a0.png)

**Comment by leiserfg** [comment]
So looks like   `!=`  `/=`  and `=/=` are incorrect.

**Comment by kovidgoyal** [comment]
Those look like bugs in Iosevka, not kitty. For instance:

```
hb-shape --show-extents --cluster-level=1 --shapers=ot iosevka-regular.ttf '!='  
[.gid9185=0+500<369,757,262,-834>|.gid9177=1+500<-436,464,872,-248>]
```

shows there are no .join suffixes for this ligature. Contrast with
```
hb-shape --show-extents --cluster-level=1 --shapers=ot iosevka-regular.ttf '=='                     
[.gid9172.join-r=0+500<64,464,519,-248>|.gid9173.join-l=1+500<-83,464,519,-248>]
```

which has the join suffixes.

**Comment by be5invis** [comment]
@kovidgoyal could you try this updated version?

[iosevka-regular.zip](https://github.com/be5invis/Iosevka/files/6595544/iosevka-regular.zip)

**Comment by be5invis** [comment]
For sample, this is under `calt`:
![image](https://user-images.githubusercontent.com/240091/120741105-47ea8100-c4a9-11eb-9a33-090d8a16fb2b.png)

This is under `dlig`:
![image](https://user-images.githubusercontent.com/240091/120741405-db23b680-c4a9-11eb-9df8-1c64974fb4a1.png)

**Comment by be5invis** [comment]
This is a further updated with some fixes on arrowheads:
[iosevka-regular.zip](https://github.com/be5invis/Iosevka/files/6595592/iosevka-regular.zip)

Test corpus:
```
-<< -< -<- <-- <--- <<- <- -> ->> --> ---> ->- >- >>-
=<< =< =<= <== <=== <<= <= => =>> ==> ===> =>= >= >>=
<-> <--> <---> <----> <=> <==> <===> <====> -------->
<~~ <~ ~> ~~> :: ::: == != /= ~= <> === !== =/= =!=
:= :- :+ <* <*> *> <| <|> |> <. <.> .> +: -: =: :> __
(* *) [| |] {| |} ++ +++ \/ /\ |- -| <!-- <!--- <***>
˩˥ ˧˥ ˩˦˩˧ ꜒꜒ ꜒꜓ ꜔꜕꜓꜖ =>>= =<<= <-|=> << <<< >> >>>
```

`calt` reference:
![image](https://user-images.githubusercontent.com/240091/120742207-5a65ba00-c4ab-11eb-910b-91d1083ff6b5.png)

`dlig` reference:
![image](https://user-images.githubusercontent.com/240091/120742229-63568b80-c4ab-11eb-9575-33b47d82bda2.png)

**Comment by kovidgoyal** [comment]
Looks OK to me now.

![ligs](https://user-images.githubusercontent.com/1308621/120748876-1d3c0f00-c521-11eb-9510-6d54e6f0c0e2.png)

**Comment by kovidgoyal** [comment]
Actually /= is still wrong
```
hb-shape --show-extents --cluster-level=1 --shapers=ot iosevka-regular.ttf '/='                   
[u002F=0+500<59,823,382,-966>|u003D=1+500<64,464,372,-248>]
```

**Comment by kovidgoyal** [comment]
And same for =/=
```
hb-shape --show-extents --cluster-level=1 --shapers=ot iosevka-regular.ttf '=/='                          
[u003D=0+500<64,464,372,-248>|u002F=1+500<59,823,382,-966>|u003D=2+500<64,464,372,-248>]
```

**Comment by kovidgoyal** [comment]
And =!=
```
hb-shape --show-extents --cluster-level=1 --shapers=ot iosevka-regular.ttf '=!='                          
[u003D=0+500<64,464,372,-248>|u0021=1+500<180,735,141,-743>|u003D=2+500<64,464,372,-248>]
```

**Comment by be5invis** [comment]
@kovidgoyal Well in `calt`, `/=`, `=/=` and `=!=` do not ligate. They _should_ be separated.

**Comment by be5invis** [comment]
It looks like that Kitty has problems dealing with ... slash?

**Comment by kovidgoyal** [comment]
Does dlig need to be turned on explicitly? Is it off by default? And the slash glyph is too wide to fit in the bounding box of the rest of the font, so it is truncated. kitty either truncates or resizes all glyphs that dont fit in a cell, because all cells are rendered in parallel, independently, which is what gives it its performance.

**Comment by be5invis** [comment]
@kovidgoyal `dlig` is off by default. To enable it you should explicitly turn this feature on, **and turn off `calt`** since they conflict with each other.

Iosevka's slash looks like this, which is pretty... normal. the `xMax` is 442 which fits in the unit cell.
![image](https://user-images.githubusercontent.com/240091/120749912-3d84b300-c4ba-11eb-9391-74af3deb9b99.png)

**Comment by kovidgoyal** [comment]
Fortunately kitty has the ability to turn on/off features in its config, so here is the output with calt off and dlig on
![ligs](https://user-images.githubusercontent.com/1308621/120750270-b66c2500-c523-11eb-8e49-d4f87b3cf65a.png)

Those ligatures still dont work, but other one involving [| do

**Comment by be5invis** [comment]
@kovidgoyal Well, your image matches my sample pretty well now. Except for slash -- there's definitely something wrong with it.

**Comment by leiserfg** [comment]
Yes, it should be drawn like a mirror of \ but it's not.

**Comment by be5invis** [comment]
There are some non-standard tags that enable certain ligatures, including:

`JSPT`: Enables special `===` and `!==`
![image](https://user-images.githubusercontent.com/240091/120750806-d536d100-c4bb-11eb-9092-dff6e2c792bc.png)

`HSKL`: Enables `/=` as inequal
![image](https://user-images.githubusercontent.com/240091/120750633-8c7f1800-c4bb-11eb-9d4b-a37506bd6af9.png)

`FSTA`: Enables `<>` as inequal, `===` and `=!=`
![image](https://user-images.githubusercontent.com/240091/120750736-bd5f4d00-c4bb-11eb-9220-93ab91b7589f.png)

`MTLB`: Enables `~=` as inequal
![image](https://user-images.githubusercontent.com/240091/120750883-f13a7280-c4bb-11eb-84fc-94717d3d5125.png)

**Comment by kovidgoyal** [comment]
Its not a ligature with dlig either:
```
hb-shape --show-extents --cluster-level=1 --shapers=ot --features "calt=0,dlig=1" iosevka-regular.ttf '/='
[u002F=0+500<59,823,382,-966>|u003D=1+500<64,464,372,-248>]
```

There rest of the features should work fine, kitty doesnt special case anything on font features other than calt. calt based ligatures are broken when the cursor is over them. Other kinds of ligatures wont be.

As for why / is redering differently, thats my mistake (my kitty config uses a special font for the slash character). With default config its fine.

![ligs](https://user-images.githubusercontent.com/1308621/120751510-bd943280-c525-11eb-86e6-9055be882877.png)

**Comment by kovidgoyal** [comment]
So only remaining issue is the lack of ligature glyph names for /= and similar even under dlig

**Comment by be5invis** [comment]
@kovidgoyal Well even in `dlig`, `/=` doesn't ligate too. `/=` is only ligated under non-standard tags targeting Haskell-like languages (since in these languages they really use `/=` as inequality sign).

**Comment by kovidgoyal** [comment]
OK, then as far as I am concerned, this issue can be closed. People that want those extra ligatures can turn on the features in kitty.conf

**Comment by be5invis** [comment]
@kovidgoyal Could you please kindly test the non-standard tags that I spotted in previous comments? They involve special glyphs and I'd like to validate whether they work correctly.

**Comment by kovidgoyal** [comment]
I tried jspt, doesnt seem to have any effect, as you can check for yourself, using hb-shape

```
hb-shape --show-extents --cluster-level=1 --shapers=ot --features "-calt,+dlig,+jspt" iosevka-regular.ttf '==='
[.gid9172.join-r=0+500<64,464,519,-248>|.gid9173.join-l=1+500<-83,464,519,-248>|.gid9174.join-l=2+500<-147,464,583,-248>]
hb-shape --show-extents --cluster-level=1 --shapers=ot --features "-calt,+dlig,-jspt" iosevka-regular.ttf '==='
[.gid9172.join-r=0+500<64,464,519,-248>|.gid9173.join-l=1+500<-83,464,519,-248>|.gid9174.join-l=2+500<-147,464,583,-248>]
```

There is no difference, in both cases, the dlig ligature is used. If I use "-calt,-dlig,+jspt" no ligation happens at all.

**Comment by kovidgoyal** [comment]
Ah never mind needs to be uppercased

**Comment by be5invis** [comment]
@kovidgoyal Yeah non-standard tags should be all uppercase letters.

**Comment by kovidgoyal** [comment]
Here is -calt +JSPT
![ligs](https://user-images.githubusercontent.com/1308621/120752678-b66e2400-c527-11eb-9f41-507d016d7e46.png)

**Comment by be5invis** [comment]
@kovidgoyal `JSPT` looked fine. How about `HSKL`, `FSTA` and `MTLB`?

**Comment by kovidgoyal** [comment]
kitty --config=NONE -o font_family=Iosevka -o 'font_features Iosevka -calt +HSKL' sh -c "cat /t/iosevka-corpus.txt; echo; cat"
![ligs](https://user-images.githubusercontent.com/1308621/120752936-1238ad00-c528-11eb-8f84-80f0b2b011c0.png)

**Comment by kovidgoyal** [comment]
kitty --config=NONE -o font_family=Iosevka -o 'font_features Iosevka -calt +FSKL' sh -c "cat /t/iosevka-corpus.txt; echo; cat"
![ligs](https://user-images.githubusercontent.com/1308621/120753014-31cfd580-c528-11eb-938c-b115386aa72a.png)

**Comment by kovidgoyal** [comment]
kitty --config=NONE -o font_family=Iosevka -o 'font_features Iosevka -calt +MTLB' sh -c "cat /t/iosevka-corpus.txt; echo; cat"
![ligs](https://user-images.githubusercontent.com/1308621/120753126-5461ee80-c528-11eb-9a72-027418570a40.png)

**Comment by be5invis** [comment]
@kovidgoyal Typo: <s>`FSKL`</s> → `FSTA`

**Comment by kovidgoyal** [comment]
kitty --config=NONE -o font_family=Iosevka -o 'font_features Iosevka -calt +FSTA' sh -c "cat /t/iosevka-corpus.txt; echo; cat"
![Uploading ligs.png…]()

**Comment by be5invis** [comment]
@kovidgoyal **LGTM**.
